### ISSUE TITLE
How do I tell nlmixr that some parameters are fixed?

### ISSUE THREAD

**Issue by fgibbons** [nlmixrdevelopment/nlmixr#272]
Hi,

nlmixr newbie here (finally got it installed/running, excited to start using!). I've had a look through this forum, as well as conference materials and the CPT:PSP article, but couldn't see how to tell nlmixr *not* to estimate certain parameters - I might wish to estimate later, but for right now, I just want them fixed at a certain value, while I focus on others. How can I do that? Thanks in advance.

-Frank

**Comment by mattfidler** [comment]
Hi @fgibbons 

There is some documentation [here](https://nlmixrdevelopment.github.io/nlmixr/articles/modelPiping.html#changing-and-fixing-parameter-values-in-models) but it is incomplete;  You may specify a fixed parameter by using `fix` or `fixed` for the parameter estimates.  A reproducible example is below (with the development version of nlmixr so the output is slightly different).

<details>

``` r
library(nlmixr)

one.compartment <- function() {
    ini({
        tka <- 0.45 # Log Ka
        tcl <- 1 # Log Cl
        tv <- fix(3.45)    # Log V
        eta.ka ~ 0.6
        eta.cl ~ 0.3
        eta.v ~ 0.1
        add.sd <- 0.7
    })
    model({
        ka <- exp(tka + eta.ka)
        cl <- exp(tcl + eta.cl)
        v <- exp(tv + eta.v)
        d/dt(depot) = -ka * depot
        d/dt(center) = ka * depot - cl / v * center
        cp = center / v
        cp ~ add(add.sd)
    })
}

fit1 <- nlmixr(one.compartment, theo_sd, est="focei",
               control=list(print=0))
#> creating full model...pruning branches...done
#> loading into symengine environment...done
#> creating full model...pruning branches...done
#> loading into symengine environment...done
#> calculate jacobian
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> calculate sensitivities
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> calculate ∂(f)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> calculate ∂(R²)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> finding duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> optimizing duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> finding duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> optimizing duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> compiling inner model...done
#> compiling EBE model...done
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> Calculating residuals/tables
#> done.
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : Initial ETAs were nudged; (Can control by foceiControl(etaNudge=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod
#> = FALSE, : ETAs were reset to zero during optimization; (Can control by
#> foceiControl(resetEtaP=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : Gradient problems with initial estimate and covariance; see $scaleInfo
print(fit1)
#> ── nlmixr FOCEi (outer: nlminb) fit ────────────────────────────────────────────
#> 
#> 
#>           OBJF      AIC      BIC Log-likelihood Condition Number
#> FOCEi 116.8533 371.4531 388.7499      -179.7265          2.19793
#> ── Time (sec $time): ───────────────────────────────────────────────────────────
#> 
#> 
#>            setup optimize covariance table    other
#> elapsed 2.259478 0.172871   0.172872 0.104 1.500779
#> ── Population Parameters ($parFixed or $parFixedDf): ───────────────────────────
#> 
#> 
#>        Parameter  Est.    SE  %RSE Back-transformed(95%CI) BSV(CV%) Shrink(SD)%
#> tka       Log Ka 0.467 0.212  45.5        1.6 (1.05, 2.42)     71.7      2.88% 
#> tcl       Log Cl  1.01 0.315  31.1       2.76 (1.49, 5.11)     26.7      3.75% 
#> tv         Log V  3.45 FIXED FIXED                    31.5     13.9      10.6% 
#> add.sd           0.695                               0.695                     
#>  
#>   Covariance Type ($covMethod): r,s
#>   Fixed parameter correlations in $cor
#>   No correlations in between subject variability (BSV) matrix
#>   Full BSV covariance ($omega) or correlation ($omegaR; diagonals=SDs) 
#>   Distribution stats (mean/skewness/kurtosis/p-value) available in $shrink 
#>   Minimization message ($message):  
#>     false convergence (8) 
#>   In an ODE system, false convergence may mean "useless" evaluations were performed.
#>   See https://tinyurl.com/yyrrwkce
#>   It could also mean the convergence is poor, check results before accepting fit
#>   You may also try a good derivative free optimization:
#>     nlmixr(...,control=list(outerOpt="bobyqa"))
#> ── Fit Data (object is a modified tibble): ─────────────────────────────────────
#> 
#> # A tibble: 132 x 22
#>   ID     TIME    DV  EVID  PRED    RES   WRES IPRED   IRES  IWRES CPRED   CRES
#>   <fct> <dbl> <dbl> <int> <dbl>  <dbl>  <dbl> <dbl>  <dbl>  <dbl> <dbl>  <dbl>
#> 1 1     0      0.74     0  0     0.74   0.391  0     0.74   1.06   0     0.74 
#> 2 1     0.25   2.84     0  3.30 -0.463 -0.245  3.84 -1.00  -1.45   3.27 -0.426
#> 3 1     0.570  6.57     0  5.90  0.674  0.356  6.79 -0.216 -0.310  5.85  0.719
#> # … with 129 more rows, and 10 more variables: CWRES <dbl>, eta.ka <dbl>,
#> #   eta.cl <dbl>, eta.v <dbl>, ka <dbl>, cl <dbl>, v <dbl>, cp <dbl>,
#> #   depot <dbl>, center <dbl>


one.compartment <- function() {
    ini({
        tka <- 0.45 # Log Ka
        tcl <- 1 # Log Cl
        tv <- 3.45    # Log V
        eta.ka ~ 0.6
        eta.cl ~ 0.3
        eta.v ~ fix(0.1)
        add.sd <- 0.7
    })
    model({
        ka <- exp(tka + eta.ka)
        cl <- exp(tcl + eta.cl)
        v <- exp(tv + eta.v)
        d/dt(depot) = -ka * depot
        d/dt(center) = ka * depot - cl / v * center
        cp = center / v
        cp ~ add(add.sd)
    })
}


fit2 <- nlmixr(one.compartment, theo_sd, est="focei",
               control=list(print=0))
#> creating full model...pruning branches...done
#> loading into symengine environment...done
#> creating full model...pruning branches...done
#> loading into symengine environment...done
#> calculate jacobian
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> calculate sensitivities
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> calculate ∂(f)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> calculate ∂(R²)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> finding duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> optimizing duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> finding duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> optimizing duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> compiling inner model...done
#> compiling EBE model...done
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> Calculating residuals/tables
#> done.
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : Initial ETAs were nudged; (Can control by foceiControl(etaNudge=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod
#> = FALSE, : ETAs were reset to zero during optimization; (Can control by
#> foceiControl(resetEtaP=.))
#> Warning in (function (uif, data, est = NULL, control = list(), ..., sum.prod =
#> FALSE, : Gradient problems with initial estimate and covariance; see $scaleInfo
print(fit2)
#> ── nlmixr FOCEi (outer: nlminb) fit ────────────────────────────────────────────
#> 
#> 
#>           OBJF      AIC      BIC Log-likelihood Condition Number
#> FOCEi 124.3883 378.9881 396.2849      -183.4941         56.72787
#> ── Time (sec $time): ───────────────────────────────────────────────────────────
#> 
#> 
#>            setup optimize covariance table   other
#> elapsed 1.415175 0.301037   0.301038  0.02 0.68175
#> ── Population Parameters ($parFixed or $parFixedDf): ───────────────────────────
#> 
#> 
#>        Parameter  Est.     SE %RSE Back-transformed(95%CI)  BSV(CV%)
#> tka       Log Ka 0.479   0.18 37.5        1.61 (1.13, 2.3)      72.3
#> tcl       Log Cl  1.01 0.0707 7.04       2.73 (2.38, 3.14)      23.7
#> tv         Log V  3.47 0.0436 1.26       32.1 (29.4, 34.9) fix(32.4)
#> add.sd           0.695                               0.695          
#>        Shrink(SD)%
#> tka         2.03% 
#> tcl         5.47% 
#> tv          48.5% 
#> add.sd            
#>  
#>   Covariance Type ($covMethod): r,s
#>   Fixed parameter correlations in $cor
#>   No correlations in between subject variability (BSV) matrix
#>   Full BSV covariance ($omega) or correlation ($omegaR; diagonals=SDs) 
#>   Distribution stats (mean/skewness/kurtosis/p-value) available in $shrink 
#>   Minimization message ($message):  
#>     false convergence (8) 
#>   In an ODE system, false convergence may mean "useless" evaluations were performed.
#>   See https://tinyurl.com/yyrrwkce
#>   It could also mean the convergence is poor, check results before accepting fit
#>   You may also try a good derivative free optimization:
#>     nlmixr(...,control=list(outerOpt="bobyqa"))
#> ── Fit Data (object is a modified tibble): ─────────────────────────────────────
#> 
#> # A tibble: 132 x 22
#>   ID     TIME    DV  EVID  PRED    RES   WRES IPRED   IRES  IWRES CPRED   CRES
#>   <fct> <dbl> <dbl> <int> <dbl>  <dbl>  <dbl> <dbl>  <dbl>  <dbl> <dbl>  <dbl>
#> 1 1     0      0.74     0  0     0.74   0.386  0     0.74   1.06   0     0.74 
#> 2 1     0.25   2.84     0  3.28 -0.436 -0.227  3.83 -0.988 -1.42   3.23 -0.394
#> 3 1     0.570  6.57     0  5.84  0.733  0.382  6.78 -0.214 -0.308  5.77  0.796
#> # … with 129 more rows, and 10 more variables: CWRES <dbl>, eta.ka <dbl>,
#> #   eta.cl <dbl>, eta.v <dbl>, ka <dbl>, cl <dbl>, v <dbl>, cp <dbl>,
#> #   depot <dbl>, center <dbl>
```

<sup>Created on 2020-01-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

</details>

**Comment by mattfidler** [comment]
Les me know if it works for you.
